### PR TITLE
Add support for rankllama

### PR DESCRIPTION
**Pull request by aniquetahir**
This pull request adds support for RankLlama with LoRA. Detailed instructions are included in `README.md`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pecos/xmr/reranker/trainer.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+from dataclasses import dataclass
+from typing import Optional, Any, Tuple, Dict
+
+import torch
+from torch.utils.data import DataLoader
+from transformers import Trainer, TrainingArguments, HfArgumentParser
+
+import pecos
+
+TIME_FORMAT_STR: str = "%b_%d_%H_%M_%S"
```
Comment (OctoberChang) [inline pecos/xmr/reranker/trainer.py]: This seems not being used. If so, remove it.

File: pecos/xmr/reranker/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# PECOS XMR Reranker
```
Comment (OctoberChang) [inline pecos/xmr/reranker/README.md:1]: Some general comments about the README.md:
- We should also introduce the data schema for training/inference?
- For the Command Line Usage (CLI), we have `pecos.xmr.reranker.train`.  Should we also have `pecos.xmc.reranker.predict`?
- Do we want to support Python API usage?

File: pecos/xmr/reranker/model.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+        supported by this method is limited to the `peft` package.
+
+        Args:
+            model (PreTrainedModel): A PreTrainedModel from the transformers package.
+        """
+        if mod_config["modifier_type"] == "peft":
+            config_type = getattr(peft, mod_config["config_type"])
+            peft_config: PeftConfig = config_type(**mod_config["config"])
+
+            # IMPORTANT This is needed for the peft adapters to train the classification head
```
Comment (OctoberChang) [inline pecos/xmr/reranker/model.py]: Do we still need these comments (the following 4 lines)?

File: pecos/xmr/reranker/model.py [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+        train_params: TrainParams,
+    ):
+        """
+        Train the ranking model
+        Args:
+            train_dataset: The training dataset (IterableDataset)
+            params: The model parameters (RankingModelParams)
+        """
+        training_args = train_params.training_args
+        training_args.remove_unused_columns = False
```
Comment (OctoberChang) [inline pecos/xmr/reranker/model.py:592]: Is this line still necessary?

File: pecos/xmr/reranker/README.md (hunk 2 above)
Comment (OctoberChang) [reply pecos/xmr/reranker/README.md:1]: Another high level comment. Can we avoid hard-coded the input columns and make them configurable in the config JSON file? Some hard-coded columns, for example:
- Line 79 of `data_utils.py`: `keywords` 
- Line 110 of `data_utils.py`: `contents`, `titles`
- Line 296-298 of `model.py`: `inp_id`, `ret_idxs`, `rel`

File: pecos/xmr/reranker/model.py [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+        """
+        Returns the forward output of the huggingface model
+        """
+        return self.hf_model(*args, **kwargs)
+
+    def gradient_checkpointing_enable(self, **kwargs):
+        """
+        Enable gradient checkpointing for the model
+        """
+        self.hf_model.enable_input_require_grads()
```
Comment (OctoberChang) [inline pecos/xmr/reranker/model.py]: Not sure if this is the right place to call `hf_model.enable_input_require_grads()`.
From Tevatron RankLlaMA implementation (https://github.com/texttron/tevatron/blob/main/src/tevatron/reranker/modeling.py#L79), they are calling only when both "LoRA" and "training_args.gradient_checkpointing" is enable.

File: pecos/xmr/reranker/trainer.py (hunk 1 above)
Comment (aniquetahir) [reply pecos/xmr/reranker/trainer.py]: removed

File: pecos/xmr/reranker/README.md (hunk 2 above)
Comment (aniquetahir) [reply pecos/xmr/reranker/README.md:1]: > Another high level comment. Can we avoid hard-coded the input columns and make them configurable in the config JSON file? Some hard-coded columns, for example:
> 
>     * Line 79 of `data_utils.py`: `keywords`
> 
>     * Line 110 of `data_utils.py`: `contents`, `titles`
> 
>     * Line 296-298 of `model.py`: `inp_id`, `ret_idxs`, `rel`

This can now be specified in the configuration. I added details in the `README.md`.




> Some general comments about the README.md:
> 
>     * We should also introduce the data schema for training/inference?
> 
>     * For the Command Line Usage (CLI), we have `pecos.xmr.reranker.train`.  Should we also have `pecos.xmc.reranker.predict`?
> 
>     * Do we want to support Python API usage?

Added predictions.

File: pecos/xmr/reranker/model.py (hunk 3 above)
Comment (aniquetahir) [reply pecos/xmr/reranker/model.py]: removed.

File: pecos/xmr/reranker/model.py (hunk 4 above)
Comment (aniquetahir) [reply pecos/xmr/reranker/model.py:592]: Since we are adding additional information to the output of the collate function, this is needed to avoid it being removed by the trainer.

File: pecos/xmr/reranker/model.py (hunk 5 above)
Comment (aniquetahir) [reply pecos/xmr/reranker/model.py]: The following sequence of operation is expected:
1. First get_modifed_model is called.
2. the trainer calls `gradient_checkpointing_enable`, when gradient checkpointing is enabled.

File: pecos/xmr/reranker/model.py [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+    def _collate_eval(
+        cls,
+        tokenizer: Union[PreTrainedTokenizer, AutoTokenizer],
+        eval_params: EvalParams,
+        features: List[str],
+        inp_idxs: List[int],
+        lbl_idxs: List[int],
+    ):
+        """
+        Collate function for training. Tokenizes the input and return features and returns the collated batch.
```
Comment (OctoberChang) [inline pecos/xmr/reranker/model.py:484]: Doc String seems to be out-dated.

File: pecos/xmr/reranker/model.py [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+
+    @classmethod
+    def init_model(cls, model_params: ModelParams, train_params: TrainParams):
+        """Initiate a model with training parameters
+
+        Args:
+            model_params (RankingModel.ModelParams): the model parameters
+            train_params (RankingModel.TrainParams): the training parameters
+        Returns:
+            An instance of UberGlobalModel
```
Comment (OctoberChang) [inline pecos/xmr/reranker/model.py]: Doc String seems to be out-dated. Remove `UberGlobalModel`.

Comment (aniquetahir) [reply pecos/xmr/reranker/model.py]: Updated docstrings

File: setup.py [hunk 8]
```diff
@@ -119,7 +119,10 @@ def get_blas_lib_dir(cls):
     'torch>=2.0; python_version>="3.8"',
     'sentencepiece>=0.1.86,!=0.1.92', # 0.1.92 results in error for transformers
     'transformers>=4.1.1; python_version<"3.9"',
-    'transformers>=4.4.2; python_version>="3.9"'
+    'transformers>=4.4.2; python_version>="3.9"',
+    'tqdm>=4.66.4',
+    'peft>=0.11.0',
```
Comment (OctoberChang) [inline setup.py]: The `peft` package requires `python_version >= 3.8` that's why the latest unit test failed.

Please also check the minimal support python version for other libraries you introduce.

**Review by OctoberChang — APPROVED** [review]
LGTM.
